vault/acl_test.go

```diff
@@ -199,6 +213,211 @@ func testLayeredACL(t *testing.T, acl *ACL) {
+	}
+	acl, err := NewACL([]*Policy{policy})
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	type tcase struct {
+		path      string
+		parameter string
+		allowed   bool
+		rootPrivs bool
+	}
+
+	toperations := []logical.Operation{
+		logical.UpdateOperation,
+		logical.CreateOperation,
+		logical.DeleteOperation,
+	}
+
+	tcases := []tcase{
+		{"foo/bar", "baz", false, false},
+		{"foo/bar", "zip", false, false},
+		{"hello/universe", "bob", true, false},
+		{"hello/universe", "tom", true, false},
+		{"rainy/day", "bob", true, false},
+		{"rainy/day", "tom", true, false},
+		{"cool/bike", "four", false, false},
+		{"cool/bike", "frank", false, false},
+		{"clean/bed", "one", false, false},
+		{"clean/bed", "two", false, false},
+		{"coca/cola", "john", false, false},
+		{"coca/cola", "two", false, false},
+	}
+
+	for _, tc := range tcases {
+		request := logical.Request{Path: tc.path, Data: make(map[string]interface{})}
+		request.Data[tc.parameter] = ""
+		for _, op := range toperations {
+			request.Operation = op
+			allowed, rootPrivs := acl.AllowOperation(&request)
+			if allowed != tc.allowed {
+				t.Fatalf("bad: case %#v: %v, %v", tc, allowed, rootPrivs)
+			}
+			if rootPrivs != tc.rootPrivs {
+				t.Fatalf("bad: case %#v: %v, %v", tc, allowed, rootPrivs)
+			}
+		}
+	}
+}
+func TestAllowOperation(t *testing.T) {
+  policy, err := Parse(permissionsPolicy)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	acl, err := NewACL([]*Policy{policy})
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+  toperations := []logical.Operation {
+    logical.UpdateOperation,
+    logical.DeleteOperation,
+    logical.CreateOperation,
+  }
+	type tcase struct {
+		path  string
+		parameter string
+		allowed   bool
+		rootPrivs bool
+	}
+	
+	tcases := []tcase{
+		{"dev/ops", "zip", true, false},
+		{"foo/bar", "zap", false, false},
+		{"foo/baz", "hello", true, false},
+		{"foo/baz", "zap", false, false},
+		{"broken/phone", "steve", false, false},
+		{"hello/world", "one", false, false},
+		{"tree/fort", "one", true, false},
+		{"tree/fort", "beer", false, false},
+		{"fruit/apple", "pear", false, false},
+		{"fruit/apple", "one", false, false},
+		{"cold/weather", "four", true, false},
+  }
+  
+  for _, tc := range tcases {
+		request := logical.Request{Path: tc.path, Data: make(map[string]interface{})}
+		request.Data[tc.parameter] = ""
+		for _, op := range toperations {
+			request.Operation = op
+			allowed, rootPrivs := acl.AllowOperation(&request)
+			if allowed != tc.allowed {
+				t.Fatalf("bad: case %#v: %v, %v", tc, allowed, rootPrivs)
+			}
+			if rootPrivs != tc.rootPrivs {
+				t.Fatalf("bad: case %#v: %v, %v", tc, allowed, rootPrivs)
+			}
+		}
+	}
+}
+
+//test merging
+
+var permissionsPolicy2 = `
+name = "ops"
+path "foo/bar" {
+	policy = "write"
+	permissions = {
+		deniedparameters = {
+			"baz" = {}
+		}
+	}
+}
+path "foo/bar" {
+	policy = "write"
+	permissions = {
+		deniedparameters = {
+			"zip" = {}
```

Comment:
These shouldn't be maps, they should be arrays:
```
path "foo/bar" {
 	policy = "write"
 	permissions = {
 		deniedparameters = {
 			"zip" = []
```
The type inside the code that this should map to is []interface{}.